### PR TITLE
Rework `LoadSpec`

### LINKED ISSUE #134: Rework `LoadSpec`
## Use Case

As a developer, the `LoadSpec` type doesn't capture all settings that are relevant when loading a volume.

## Solution

- Remove the `subpath` field, replace with something like `resolutionLevel`
- Consider giving extent fields in pixels at provided resolution level, rather than 0-1 floats
- Add a list of channel indexes to load

### PR DESCRIPTION
**Pull request by frasercl**
Fully resolves #134: reworks `LoadSpec` to have all the fields we want and none of the fields we don't.

- Removed field: `subpath`. No one wanted to use it.
- Added optional field: `multiscaleLevel`. This specifies which scale level to load (in zarr only). The loader will still pick the optimal level if `multiscaleLevel` is undefined or larger than the texture limits, so this is more like a "minimum downsampling level."
- Added optional field: `channels`. This lets you pick a subset of channel indexes to load.

Also, updated loaders to always set `Volume.loadSpec` to reflect the data that was *actually* loaded, rather than the `LoadSpec` that was passed in. E.g. no loader other than zarr can use `subregion`, so keep it set to the max. And since json loader has to load channels in static groups of (up to) four, when the user requests some subset of channels, also include all the channels that came along with the requested channel's images.

### CODE REVIEW

**Review by ShrimpCryptid — COMMENTED** [review]
Could you talk a bit more about why subpath was removed?

File: src/Volume.ts [hunk 1]
```diff
@@ -130,7 +130,7 @@ export default class Volume {
   public imageInfo: ImageInfo;
   public loadSpec: LoadSpec;
   public loader?: IVolumeLoader;
-  private loadSpecRequired: LoadSpec;
+  private loadSpecRequired: Required<LoadSpec>;
```
Comment (toloudis) [inline src/Volume.ts:133]: What do you think of a comment up here saying why loadSpecRequired is different than loadSpec?  Just to show the intention...

File: src/loaders/JsonImageInfoLoader.ts [hunk 2]
```diff
@@ -231,60 +249,62 @@ class JsonImageInfoLoader implements IVolumeLoader {
...
-          for (let ch = 0; ch < Math.min(thisbatch.length, 4); ++ch) {
-            channelsBits.push(new Uint8Array(w * h));
-          }
-          // extract the data
-          for (let j = 0; j < Math.min(thisbatch.length, 4); ++j) {
-            for (let px = 0; px < w * h; px++) {
-              channelsBits[j][px] = iData.data[px * 4 + j];
-            }
+
+      img.onload = (event) => {
```
Comment (toloudis) [inline src/loaders/JsonImageInfoLoader.ts:257]: This is a big improvement!

File: src/loaders/OpenCellLoader.ts [hunk 3]
```diff
@@ -48,13 +48,14 @@ class OpenCellLoader implements IVolumeLoader {
     };
 
     // got some data, now let's construct the volume.
-    const vol = new Volume(imgdata, loadSpec, this);
+    // This loader uses no fields from `LoadSpec`. Iinitialize volume with defaults.
```
Comment (toloudis) [inline src/loaders/OpenCellLoader.ts:51]:
```suggestion
    // This loader uses no fields from `LoadSpec`. Initialize volume with defaults.
```

**Review by toloudis — APPROVED** [review]
Couple small comments but the functional stuff looks good to me!

File: src/loaders/JsonImageInfoLoader.ts (hunk 2 above)
Comment (frasercl) [reply src/loaders/JsonImageInfoLoader.ts:257]: I (mercifully) never had a reason to use that make-a-function-that-returns-a-function-then-immediately-call-the-outer-function hack in the era before arrow functions, and I don't envy folks who had to write software professionally back when that was an accepted solution.

**Comment by frasercl** [comment]
> Could you talk a bit more about why subpath was removed?

As I understand it, the only thing `subpath` might have been used for is what `multiscaleLevel` does now: picking different multiscale levels, by name rather than index (if they were different - I think that was rare anyways). It was never really used anywhere and didn't have a clear reason to keep existing. @toloudis might know more though.

**Comment by toloudis** [comment]
> > Could you talk a bit more about why subpath was removed?
> 
> As I understand it, the only thing `subpath` might have been used for is what `multiscaleLevel` does now: picking different multiscale levels, by name rather than index (if they were different - I think that was rare anyways). It was never really used anywhere and didn't have a clear reason to keep existing. @toloudis might know more though.

Yes - in zarr metadata, a multiscale level is really pointing to a folder that contains zarr arrays. And in practice, often the folder name is the same as the integer index.   So subpath (aka multiscale level as string) was an early way to make it work but it's only relevant for Zarr.  An integer level index is much more sensible and general - and in the Zarr case, it is used to find the subpath to the zarr data for that level.

**Review by ShrimpCryptid — APPROVED** [review]
Thanks for clarifying, looks good!

**Pull request merged**